### PR TITLE
Prep work for Script Compilation: Factoring

### PR DESCRIPTION
**Pull request by vpax**
The commits in this PR factor out existing functionality into new functions/methods (which ultimately script compilation will use directly). The most significant change is allowing derivation of `ScriptFunc` subclasses, in particular in order to deal with lambda captures in ways differing from the current functionality.  In the process of adding that, I found what appears to be buggy functionality in serializing lambdas via Broker.  I didn't try to fix that, but added a comment to flag it.

### CODE REVIEW

File: src/Expr.h
Comment (timwoj) [inline src/Expr.h:942]: Could you add some comments for these?

File: src/Expr.cc
Comment (timwoj) [inline src/Expr.cc:3907]: Unused variable?

File: src/Expr.cc
Comment (timwoj) [inline src/Expr.cc:3796]: This block could be `std::fill_n(map.begin(), map_size, -1)`.

File: src/Expr.h
Comment (timwoj) [inline src/Expr.h:1186]: Does this need to be in the global namespace?

File: src/Func.h
Comment (timwoj) [inline src/Func.h:216]: I think a better name is `HasCopySemantics()` since it's more clear that it's checking for something.

**Review by timwoj — CHANGES_REQUESTED** [review]

File: src/Expr.h
Comment (vpax) [reply src/Expr.h:942]: roger, will be in the next push

File: src/Expr.cc
Comment (vpax) [reply src/Expr.cc:3907]: It's the index for the loop here: https://github.com/zeek/zeek/blob/9a65bb4cd0be45ac8d95206c85516be333ae2001/src/Expr.cc#L3913

File: src/Expr.cc
Comment (vpax) [reply src/Expr.cc:3796]: Nifty, didn't know about that.

File: src/Expr.h
Comment (vpax) [reply src/Expr.h:1186]: The compiler needs to generate code that'll call it.  I guess it could be a static member function, though for my taste that seems a bit weird since it would tie its definition to the interpreter when it's really meant to be stand-alone functionality.

File: src/Func.h
Comment (vpax) [reply src/Func.h:216]: Agreed.

File: src/Expr.cc
Comment (vpax) [reply src/Expr.cc:3796]: Actually, turns out that to use this requires a preceding `resize()`, and that allows this particular operation directly.

**Comment by timwoj** [comment]
After merging the ZVal work, there's now an include loop from Expr.h -> Val.h -> ZVal.h -> Expr.h and the build breaks. Can you rebase this on master and fix it? The changes look fine otherwise and I'll merge it after that.

**Comment by vpax** [comment]
@timwoj I think this is now fixed, see what you think.

**Pull request merged**
